File: Crio/crio_programming_interview_problems-master/HappyNumber/HappyNumber.cpp
#include <bits/stdc++.h>

using namespace std;

class HappyNumber {
    public:
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        int sq(int n) {
            int res = 0;
            while(n>0) {
                int x = n%10;
                res += x*x;
                n/=10;
            }
            return res;
        }
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

        string happyNumber(int n) {
            // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
            n = abs(n);
            map<int,int> mp;
            while(!mp[n] && n!=1) {
                mp[n]=1;
                n = sq(n);
            }
            if(n==1) return "true";
            return "false";
            // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
        }
};

int main() {
    int n;
    cin >> n;
    string result = HappyNumber().happyNumber(n);
    cout << result;
    return 0;
}

File: Crio/crio_programming_interview_problems-master/HappyNumber/HappyNumber.py
def happyNumber(n):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    vis = set()
    n= abs(n)
    while(n!=1 and n not in vis):
        vis.add(n)
        x = n
        n = 0
        while(x>0):
            n+=(x%10)**2
            x//=10

    if(n==1):
        return "true"
    return "false"
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS


if __name__ == '__main__':
    n = int(input())
    ans = happyNumber(n)
    print(ans)
